### PR TITLE
CO-31: Jenkins Integration

### PR DESCRIPTION
**Pull request by carlpulley**
- [x] Add in Jenkinsfile
- [x] Remove .drone.yml file
- [x] Ensure code builds and publishes locally
---
- [x] Ready for Review

### CODE REVIEW

**Comment by fommil** [comment]
Nice. Is somebody going to wire this up so we get CI on PR? Ani wanted to use drone but I just want something running at this point.

**Comment by carlpulley** [comment]
This file is already up and running (i.e. Jenkins builds are being triggered - see http://jenkins.cakesolutions.eu/job/cakesolutions/job/sbt-cake/). Git integration with Jenkins is likely still required?

Scripted plugin tests are currently failing though (both locally in dev environment and in CI).

**Comment by carlpulley** [comment]
~~Retest this please~~ this command doesn't appear to work in the current CI environment?

**Comment by carlpulley** [comment]
OK, it appears that the scripted tests here are failing because there's an assumption that `scalafmt` should be installed locally (e.g. within the Jenkins environment)? This suggestion is based on the console logging of http://jenkins.cakesolutions.eu/job/cakesolutions/job/sbt-cake/job/feature%252Fco-31-jenkinsfile/2/console

**Comment by carlpulley** [comment]
Hmm, in my local dev environment I have an installation of scalafmt under /usr/local/bin. In this environment I'm observing the following exception stack trace:
```
[info] java.lang.NoClassDefFoundError: Could not initialize class scala.meta.package$
[info] 	at org.scalafmt.Scalafmt$.format(Scalafmt.scala:47)
[info] 	at org.scalafmt.cli.Cli$.org$scalafmt$cli$Cli$$handleFile(Cli.scala:144)
[info] 	at org.scalafmt.cli.Cli$$anonfun$runFormat$1.apply(Cli.scala:220)
[info] 	at org.scalafmt.cli.Cli$$anonfun$runFormat$1.apply(Cli.scala:218)
[info] 	at scala.collection.Iterator$class.foreach(Iterator.scala:891)
[info] 	at scala.collection.AbstractIterator.foreach(Iterator.scala:1334)
[info] 	at scala.collection.parallel.ParIterableLike$Foreach.leaf(ParIterableLike.scala:972)
[info] 	at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply$mcV$sp(Tasks.scala:49)
[info] 	at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply(Tasks.scala:48)
[info] 	at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply(Tasks.scala:48)
[info] 	at scala.collection.parallel.Task$class.tryLeaf(Tasks.scala:51)
[info] 	at scala.collection.parallel.ParIterableLike$Foreach.tryLeaf(ParIterableLike.scala:969)
[info] 	at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask$class.internal(Tasks.scala:159)
[info] 	at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.internal(Tasks.scala:443)
[info] 	at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask$class.compute(Tasks.scala:149)
[info] 	at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.compute(Tasks.scala:443)
[info] 	at scala.concurrent.forkjoin.RecursiveAction.exec(RecursiveAction.java:160)
[info] 	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
[info] 	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
[info] 	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
[info] 	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
[info] [error] java.lang.NoClassDefFoundError: Could not initialize class scala.meta.package$
[info] [error] Use 'last' for the full log.
[error] fatal: Not a git repository (or any of the parent directories): .git
[info] [info] Compiling 1 Scala source to /private/var/folders/1d/33rn_6zs2jq2tkx6g621p6z40000gn/T/sbt_9bcd4c7e/simple/target/scala-2.11/classes...
[info] [info] Compiling 1 Scala source to /private/var/folders/1d/33rn_6zs2jq2tkx6g621p6z40000gn/T/sbt_9bcd4c7e/simple/logging/target/scala-2.11/classes...
[info] [success] Total time: 4 s, completed 26-May-2017 16:32:29
[info] java.lang.NoClassDefFoundError: Could not initialize class scala.meta.package$
[info] 	at org.scalafmt.Scalafmt$.format(Scalafmt.scala:47)
[info] 	at org.scalafmt.cli.Cli$.org$scalafmt$cli$Cli$$handleFile(Cli.scala:144)
[info] 	at org.scalafmt.cli.Cli$$anonfun$runFormat$1.apply(Cli.scala:220)
[info] 	at org.scalafmt.cli.Cli$$anonfun$runFormat$1.apply(Cli.scala:218)
[info] 	at scala.collection.Iterator$class.foreach(Iterator.scala:891)
[info] 	at scala.collection.AbstractIterator.foreach(Iterator.scala:1334)
[info] 	at scala.collection.parallel.ParIterableLike$Foreach.leaf(ParIterableLike.scala:972)
[info] 	at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply$mcV$sp(Tasks.scala:49)
[info] 	at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply(Tasks.scala:48)
[info] 	at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply(Tasks.scala:48)
[info] 	at scala.collection.parallel.Task$class.tryLeaf(Tasks.scala:51)
[info] 	at scala.collection.parallel.ParIterableLike$Foreach.tryLeaf(ParIterableLike.scala:969)
[info] 	at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask$class.internal(Tasks.scala:159)
[info] 	at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.internal(Tasks.scala:443)
[info] 	at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask$class.compute(Tasks.scala:149)
[info] 	at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.compute(Tasks.scala:443)
[info] 	at scala.concurrent.forkjoin.RecursiveAction.exec(RecursiveAction.java:160)
[info] 	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
[info] 	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
[info] 	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
[info] 	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
[info] [error] java.lang.NoClassDefFoundError: Could not initialize class scala.meta.package$
[info] [error] Use 'last' for the full log.
[error] x sbt-cake / simple
[error]    {line 7}  Command failed: scalafmt failed
Running target / resolution-cache
[error] x target / resolution-cache
java.io.FileNotFoundException: /var/folders/1d/33rn_6zs2jq2tkx6g621p6z40000gn/T/sbt_8c3248e9/resolution-cache/test (No such file or directory)
	at java.io.FileInputStream.open0(Native Method)
	at java.io.FileInputStream.open(FileInputStream.java:195)
	at java.io.FileInputStream.<init>(FileInputStream.java:138)
	at sbt.Using$$anonfun$fileInputStream$1.apply(Using.scala:73)
	at sbt.Using$$anonfun$fileInputStream$1.apply(Using.scala:73)
	at sbt.Using$$anon$2.openImpl(Using.scala:65)
	at sbt.OpenFile$class.open(Using.scala:43)
	at sbt.Using$$anon$2.open(Using.scala:64)
	at sbt.Using$$anon$2.open(Using.scala:64)
	at sbt.Using.apply(Using.scala:23)
	at sbt.IO$.transfer(IO.scala:319)
	at sbt.IO$.read(IO.scala:702)
	at xsbt.test.TestScriptParser.parse(TestScriptParser.scala:42)
	at sbt.test.ScriptedTests.runTest$1(ScriptedTests.scala:73)
	at sbt.test.ScriptedTests.sbt$test$ScriptedTests$$scriptedTest(ScriptedTests.scala:82)
	at sbt.test.ScriptedTests$$anonfun$scriptedTest$1$$anonfun$apply$1$$anonfun$apply$2$$anonfun$apply$3.apply(ScriptedTests.scala:42)
	at sbt.test.ScriptedTests$$anonfun$scriptedTest$1$$anonfun$apply$1$$anonfun$apply$2$$anonfun$apply$3.apply(ScriptedTests.scala:36)
	at sbt.Resources.sbt$Resources$$readWrite$1(Resources.scala:58)
	at sbt.Resources$$anonfun$readWriteResourceDirectory$1.apply(Resources.scala:60)
	at sbt.Resources$$anonfun$readWriteResourceDirectory$1.apply(Resources.scala:60)
	at sbt.IO$.withTemporaryDirectory(IO.scala:358)
	at sbt.Resources.readWriteResourceDirectory(Resources.scala:60)
	at sbt.Resources.readWriteResourceDirectory(Resources.scala:48)
	at sbt.test.ScriptedTests$$anonfun$scriptedTest$1$$anonfun$apply$1$$anonfun$apply$2.apply(ScriptedTests.scala:36)
	at sbt.test.ScriptedTests$$anonfun$scriptedTest$1$$anonfun$apply$1$$anonfun$apply$2.apply(ScriptedTests.scala:34)
	at sbt.test.ScriptedRunner$$anonfun$3.apply(ScriptedTests.scala:154)
	at sbt.test.ScriptedRunner$$anonfun$3.apply(ScriptedTests.scala:154)
	at scala.collection.TraversableLike$$anonfun$flatMap$1.apply(TraversableLike.scala:251)
	at scala.collection.TraversableLike$$anonfun$flatMap$1.apply(TraversableLike.scala:251)
	at scala.collection.mutable.ArraySeq.foreach(ArraySeq.scala:73)
	at scala.collection.TraversableLike$class.flatMap(TraversableLike.scala:251)
	at scala.collection.AbstractTraversable.flatMap(Traversable.scala:105)
	at sbt.test.ScriptedRunner.runAll(ScriptedTests.scala:154)
	at sbt.test.ScriptedRunner.run(ScriptedTests.scala:151)
	at sbt.test.ScriptedRunner.run(ScriptedTests.scala:125)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at sbt.ScriptedPlugin$$anonfun$scriptedTask$1$$anonfun$apply$6$$anonfun$apply$7.apply(ScriptedPlugin.scala:92)
	at sbt.ScriptedPlugin$$anonfun$scriptedTask$1$$anonfun$apply$6$$anonfun$apply$7.apply(ScriptedPlugin.scala:88)
	at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
	at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
	at sbt.std.Transform$$anon$4.work(System.scala:63)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
	at sbt.Execute.work(Execute.scala:237)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
[error] (*:scripted) java.io.FileNotFoundException: /var/folders/1d/33rn_6zs2jq2tkx6g621p6z40000gn/T/sbt_8c3248e9/resolution-cache/test (No such file or directory)
[error] Total time: 23 s, completed 26-May-2017 16:32:29
```

This dump suggests that scalafmt is found but then fails for other reasons?

**Comment by pzapletal** [comment]
@carlpulley https://github.com/scalameta/scalafmt/issues/943

**Comment by olafurpg** [comment]
I just published scalafmt 1.0.0-RC1, please give it a try and let me know if it's fixes the issue

**Comment by fommil** [comment]
bump sbt-sensible to 1.2.0 if you want to use scalafmt.

File: Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,38 @@
+pipeline {
+  agent {
+    label 'sbt-slave'
+  }
+
+  environment {
+    // Ensure that build scripts recognise the environment they are running within
+    CI = ''
+    SCALA_VERSION = '2.10.6'
```
Comment (fommil) [inline Jenkinsfile]: I don't like allowing a different version of scala here than in the `build.sbt`. This isn't a library that needs to be crossbuilt, so this only causes duplication.

Comment (fommil) [reply Jenkinsfile]: are you sure `CI` is not set by jenkins already?

**Review by fommil — CHANGES_REQUESTED** [review]

**Comment by amirkarimi** [comment]
I can confirm that scalafmt 1.0.0-RC1 runs without any exception but I get the following error when running `sbt scripted`:

```
[info] [success] Total time: 2 s, completed 30-May-2017 14:52:39
[info] All files are formatted with scalafmt :)
[error] x sbt-cake / simple
[error]    {line 2}  Command succeeded but failure was expected
java.lang.RuntimeException: Failed tests:
	sbt-cake / simple

	at scala.sys.package$.error(package.scala:27)
	at sbt.test.ScriptedRunner.runAll(ScriptedTests.scala:156)
	at sbt.test.ScriptedRunner.run(ScriptedTests.scala:151)
	at sbt.test.ScriptedRunner.run(ScriptedTests.scala:125)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at sbt.ScriptedPlugin$$anonfun$scriptedTask$1$$anonfun$apply$6$$anonfun$apply$7.apply(ScriptedPlugin.scala:92)
	at sbt.ScriptedPlugin$$anonfun$scriptedTask$1$$anonfun$apply$6$$anonfun$apply$7.apply(ScriptedPlugin.scala:88)
	at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
	at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
	at sbt.std.Transform$$anon$4.work(System.scala:63)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
	at sbt.Execute.work(Execute.scala:237)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
[error] (*:scripted) Failed tests:
[error] 	sbt-cake / simple
```

**Comment by fommil** [comment]
I already fixed scalafmt in master.

File: Jenkinsfile (hunk 1 above)
Comment (amirkarimi) [reply Jenkinsfile]: @fommil We checked that. Jenkins do not add an environment variable specifically for this purpose.

Comment (amirkarimi) [reply Jenkinsfile]: @carlpulley Isn't better to be `CI = true`?

Comment (fommil) [reply Jenkinsfile]: CI=jenkins would be better.

Comment (carlpulley) [reply Jenkinsfile]: Will add in a `jenkins` value for the setting.

`SCALA_VERSION` here refers to the version of Scala that SBT will run. Without this setting, we will run SBT with Scala 2.10.6 (regardless of this being set to another value within the project build.sbt). I'd rather be explicit regarding the version of Scala that SBT uses - particularly on a build server.

Comment (fommil) [reply Jenkinsfile]: I disagree. The internal version of scala for the build tool is an implementation detail and all we're doing here is introducing complexity and duplication for no good reason (it also will not scale for cross builds to sbt 1.0)

**Comment by carlpulley** [comment]
Able to confirm that `sbt scripted` is now passing locally. However, we still have failures in Jenkins - e.g. see http://jenkins.cakesolutions.eu/job/cakesolutions/job/sbt-cake/job/feature%252Fco-31-jenkinsfile/5/console

**Comment by carlpulley** [comment]
OK, I'm able to replicate the Jenkins failure locally (we just need to set the `CI` environment variable).

**Comment by fommil** [comment]
looks like we need `"com.google.code.findbugs" % "jsr305" % "3.0.2" % "provided"`

**Comment by carlpulley** [comment]
Issue I'm observing is down to warnings being elevated to error when `CI` is set (I can now investigate this issue further). Both `sbt-cake` and `sbt-sensible` use `CI` and so we probably need to ask if this is sensible?

@fommil why do we currently use `sbt-sensible`? Is it just for versioned releases or for other reasons?

**Comment by fommil** [comment]
to be able to release. It's not passed to any cake projects.

**Review by fommil — APPROVED** [review]

File: Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  environment {
+    // Ensure that build scripts recognise the environment they are running within
+    CI = 'jenkins'
+  }
+
+  stages {
+    stage('Compile') {
+      steps {
+        ansiColor('xterm') {
+          sh "sbt clean compile doc"
```
Comment (lashford) [inline Jenkinsfile]: run consecutivley `sbt clean; compile; doc;`

**Review by tabdulradi — APPROVED** [review]

File: Jenkinsfile (hunk 2 above)
Comment (fommil) [reply Jenkinsfile]: `;` **before** a command schedules all tasks to the sbt graph scheduler (i.e. in parallel). Spaces between commands means to run consecutively.

Comment (fommil) [reply Jenkinsfile]: you can `sbt 'clean' ';compile ;doc'` to get maximum parallelisation here.

**Pull request merged**